README.rst
===============
vmod_vslp
===============

------------------------------------------
Varnish Consistent Hashing Director Module
------------------------------------------

:Author: Julian Wiesener
:Date: 2014-07-16
:Version: 1.1
:Manual section: 3

.. _synopsis:

SYNOPSIS
========

::

	import vslp;

        sub vcl_init {
                new vd = vslp.vslp();
                vd.add_backend(be1);
                vd.add_backend(be2);
                vd.add_backend(be3);
                vd.init_hashcircle(150); #number of replicas
        }

	sub vcl_recv {
               set req.backend_hint = vd.backend();
        }

	# and/or

	sub vcl_backend_fetch {
               set bereq.backend = vd.backend();
        }


DESCRIPTION
===========

VSLP Director - Varnish StateLess Persistence Director

The purpose of this director is to provide stateless persistence (or
persistency if you prefer) for backend requests based upon this simple
concept:

* Generate a load balancing key, which will be used to select the backend
  The key values should be as uniformly distributed as possible.
  For all requests which need to hit the same backend server, the same
  key must be generated.
  For strings, a hash function can be used to generate the key.

* Select the preferred backend server using an implementation of consistent
  hashing (cf. Karger et al, references below), which ensures that the same
  hosts are always chosen for every key (for instance hash of incoming URL)
  in the same order (i.e. if the preferred host is down, then alternative
  hosts are always chosen in a fixed order).

* Initialize a circular data structure consisting of the hash values of
  "hostname%d" for each host and for a running number from 1 to n (n is
  the number of "replicas").

* For the load balancing key, find the smallest hash value in the circle
  that is larger than the key (searching clockwise and wrapping around
  as necessary).

* If the host thus selected is down, choose alternative hosts by
  continuing to search clockwise in the circle.

On consistent hashing see:

* http://www8.org/w8-papers/2a-webserver/caching/paper2.html
* http://www.audioscrobbler.net/development/ketama/
* svn://svn.audioscrobbler.net/misc/ketama
* http://en.wikipedia.org/wiki/Consistent_hashing

This technique allowes to implement persistence without keeping any state,
and, in particular, without the need to synchronize state between nodes of a
cluster of Varnish servers.

Still, all members of a cluster will, if otherwise configured equally,
implement the same persistence, in other words, requests sharing a common
persistence criterium will be balanced onto the same backend server no matter
which Varnish server handles the request.

The drawbacks are:

* the distribution of requests depends on the number of requests per key and
  the uniformity of the distribution of key values. In short, this technique
  will generally lead to less good load balancing compared to stateful
  techniques.

* When a backend server becomes unavailable, every persistence technique has
  to reselect a new backend server, but this technique will also switch back
  to the preferred server once it becomes healthy again, so the persistence
  is generally less stable compared to stateful techniques (which would
  continue to use a selected server for as long as possible (or dictated by a
  TTL)).

The name VSLP was chosen to accompany VTLA as another VETR (Varnish
Exception To the Rule) ;-) http://varnish-cache.org/wiki/VTLA



FUNCTIONS
=========

_Note_ The vmod functions differ for varnish 4.0 and 4.1. This is the
documentation for the varnish 4.1 interface.

VSLP needs to be configured before it can hand out backends. The functions not
returning a backend should only be called in vcl_init, as we can safely skip
locking for most parts of the code, if we do not modify the hash ring after
initialization.

VOID .add_backend(BACKEND)
--------------------------
Adds a backend to the Director, must be called in before init_hashcircle().

VOID .set_rampup_ratio(REAL)
----------------------------

Sets the ratio of requests (0.0 to 1.0) that goes to the next alternate backend
to warm it up when the preferd backend is healthy. If the preferred backend just
recovered from being unhealthy, the ratio will be used to warm the preferred
backend. Defaults to 0.1 (10%).

VOID .set_rampup_time(DURATION)
-------------------------------

Sets the duration that will be used to determine if a backend will be counted
as warmed after recovering from an unhealthy state. During the warm-up period,
only the rampup-ratio of requests will go to the recoverd backend. Defaults
to 60 seconds.

VOID .set_hash(ENUM { CRC32, SHA256, RS })
------------------------------------------

Sets the default hash algorithm used to choose a backend by backend(). Defaults
to CRC32.

VOID .init_hashcircle(INT)
--------------------------

Initializes the hash ring. This function must be called after all backends are
added. The argument is the numbers of replicas the hash ring contains for each
backend.

INT .hash_string(STRING string, ENUM { CRC32, SHA256, RS } alg)
---------------------------------------------------------------

Returns the hash of its first argument using the hash
algorithm defined, defaults to CRC32.


BACKEND .backend(INT nte, BOOL altsrv_p, BOOL healthy, INT hash)
----------------------------------------------------------------

Returns the nth backend with respect of altsrv_p  and respect of its healthy
state for the given hash. All parameters are optional, the defaults are:

 nte=0 will pick the first backend under respect of VSLP rules
 altsrv_p=ture
 healthy=true
 hash=0 will use a CRC32 of the request URL


LIMITATIONS
===========

* The number of backends per director is limited to 64.
* Adding backends after initializing the hash ring is possible, but invalid, as
  such backends will not be choosen. Calling init_hashcircle() twice on the same
  VSLP instance is invalid.


INSTALLATION
============

The source tree is based on autotools to configure the building, and
does also have the necessary bits in place to do functional unit tests
using the ``varnishtest`` tool.

Building requires the Varnish header files and uses pkg-config to find
the necessary paths.

Usage::

 ./autogen.sh
 ./configure

If you have installed Varnish to a non-standard directory, call
``autogen.sh`` and ``configure`` with ``PKG_CONFIG_PATH`` pointing to
the appropriate path. For example, when varnishd configure was called
with ``--prefix=$PREFIX``, use

 PKG_CONFIG_PATH=${PREFIX}/lib/pkgconfig
 export PKG_CONFIG_PATH

Make targets:

* make - builds the vmod.
* make install - installs your vmod.
* make check - runs the unit tests in ``src/tests/*.vtc``
* make distcheck - run check and prepare a tarball of the vmod.

Installation directories
------------------------

By default, the vmod ``configure`` script installs the built vmod in
the same directory as Varnish, determined via ``pkg-config(1)``. The
vmod installation directory can be overridden by passing the
``VMOD_DIR`` variable to ``configure``.

Other files like man-pages and documentation are installed in the
locations determined by ``configure``, which inherits its default
``--prefix`` setting from Varnish.


MISSING
=======
* Documentation of interactions with restarts/retries

HISTORY
=======

Version 1.0: Initial version.

ACKNOWLEDGEMENTS
================

Development of this module was partly sponsored by Deutsche Telekom AG
– Products & Innovation


COPYRIGHT
=========

This document is licensed under the same license as the
libvmod-vslp project. See LICENSE for details.

Copyright 2013-2015 UPLEX Nils Goroll Systemoptimierung. All rights reserved.


debian/changelog
libvmod-vslp (0.1-1wm) jessie-wikimedia; urgency=low

  * Local package (WMF).
  * Initial Debian packaging.

 -- Emanuele Rocca <ema@wikimedia.org>  Thu, 04 Feb 2016 17:58:20 +0100


src/vslp_dir.h
/*-
 * Copyright 2009-2013 UPLEX - Nils Goroll Systemoptimierung
 * All rights reserved.
 *
 * Author: Julian Wiesener <jw@uplex.de>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

struct vbitmap;

struct be_state {
	unsigned				magic;
#define VSLP_BESTATE_MAGIC			0x14e33bd2
	time_t					recover_time;
	uint32_t				state;
	uint32_t				canon_point;
};

struct vslp_hostnode {
	uint32_t		point;
	unsigned int		host;
};

struct vslp_host {
	VCL_BACKEND		*backend;
	uint32_t		canon_point;
};

struct vslpdir {
	unsigned				magic;
#define VSLPDIR_MAGIC				0xdbb7d59f
	pthread_mutex_t				mtx;
	unsigned				n_backend;
	unsigned				l_backend;
	VCL_BACKEND				*backend;
	struct be_state				*bstate;
	time_t					rampup_time;
	double					altsrv_p;
	hash_func				hash_fp;
	struct vslp_hostnode			*hashcircle;
	VCL_INT					replicas;
};

void vslpdir_new(struct vslpdir **vslpdp, const char *vcl_name);
void vslpdir_delete(struct vslpdir **vslpdp);
void vslpdir_add_backend(struct vslpdir *vslpd, VCL_BACKEND be);
void vslpdir_set_rampup_ratio(struct vslpdir *vslpd, VCL_REAL ratio);
void vslpdir_set_rampup_time(struct vslpdir *vslpd, VCL_DURATION duration);
void vslpdir_set_hash(struct vslpdir *vslpd, hash_func hash_m);
void vslpdir_init_hashcircle(struct vslpdir *vslpd, VCL_INT replicas);
void vslpdir_lock(struct vslpdir *vslpd);
void vslpdir_unlock(struct vslpdir *vslpd);
void vslpdir_expand(struct vslpdir *vslpd, unsigned n);
unsigned vslpdir_any_healthy(struct vslpdir *vslpd);
VCL_BACKEND vslpdir_pick_be(struct vslpdir *vslpd, const struct vrt_ctx *ctx, uint32_t hash,
	       VCL_INT n_retry, VCL_BOOL altsrv_p, VCL_BOOL healthy);


src/vslp_hash.c
/*-
 * Copyright 2009-2013 UPLEX - Nils Goroll Systemoptimierung
 * All rights reserved.
 *
 * Authors: Nils Goroll <nils.goroll@uplex.de>
 *          Geoffrey Simmons <geoff.simmons@uplex.de>
 *          Julian Wiesener <jw@uplex.de>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include "config.h"

#include <stdint.h>
#include <string.h>

#include "cache/cache.h"

#include "vrt.h"
#include "crc32.h"
#include "vsha256.h"

#include "vslp_hash.h"

uint32_t vslp_hash_crc32(VCL_STRING s)
{
	uint32_t crc;
	crc = crc32(~0U, (const unsigned char*)s, strlen(s));
	crc ^= ~0U;
	crc++;

	return (crc ? crc : 1);
}

uint32_t vslp_hash_sha256(VCL_STRING s)
{
	uint32_t res = 0;
	struct SHA256Context sha256;
	union {
		unsigned char digest[32];
		uint32_t uint32_digest[8];
	} sha256_digest;

	SHA256_Init(&sha256);
	SHA256_Update(&sha256, s, strlen(s));
	SHA256_Final(sha256_digest.digest, &sha256);

	/*
	* use low 32 bits only
	* XXX: Are these the best bits to pick?
	*/
	res = sha256_digest.uint32_digest[7];
	res++;

	return (res ? res : 1);
}

uint32_t vslp_hash_rs(VCL_STRING s)
{
	uint32_t res = 0;
	/* hash function from Robert Sedgwicks 'Algorithms in C' book */
	const uint32_t b    = 378551;
	uint32_t a          = 63689;

	while (*s) {
		res = res * a + (*s++);
		a *= b;
	}
	res++;

	return (res ? res : 1);
}

hash_func vslp_get_hash_fp(VCL_ENUM hash_m)
{
	if (!strcmp(hash_m, "CRC32"))
		return vslp_hash_crc32;
	if (!strcmp(hash_m, "SHA256"))
		return vslp_hash_sha256;
	if (!strcmp(hash_m, "RS"))
		return vslp_hash_rs;
	WRONG("Illegal VMOD enum");
}


src/vslp_hash.h
/*-
 * Copyright 2009-2013 UPLEX - Nils Goroll Systemoptimierung
 * All rights reserved.
 *
 * Author: Julian Wiesener <jw@uplex.de>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

typedef uint32_t (*hash_func)(VCL_STRING);

uint32_t vslp_hash_crc32(VCL_STRING s);
uint32_t vslp_hash_sha256(VCL_STRING s);
uint32_t vslp_hash_rs(VCL_STRING s);

hash_func vslp_get_hash_fp(VCL_ENUM hash_m);


src/vslp_dir.c
/*-
 * Copyright 2009-2015 UPLEX - Nils Goroll Systemoptimierung
 * All rights reserved.
 *
 * Authors: Nils Goroll <nils.goroll@uplex.de>
 *          Geoffrey Simmons <geoff.simmons@uplex.de>
 *          Julian Wiesener <jw@uplex.de>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include "config.h"

#include <stdlib.h>
#include <stdint.h>
#include <stdio.h>
#include <math.h>
#include <time.h>

#include "cache/cache.h"
#include "cache/cache_director.h"

#include "vrt.h"

#include "vslp_hash.h"
#include "vslp_dir.h"

typedef int (*compar)( const void*, const void* );

struct vslp_state {
	struct vslpdir *vslpd;
	uint64_t picklist;
	int idx;
	const struct vrt_ctx *ctx;
};

static int
hostnode_compare(struct vslp_hostnode *a, struct vslp_hostnode *b)
{
	return (a->point == b->point) ? 0 : ((a->point > b->point) ? 1 : -1);
}

static int
vslp_get_prefered_idx(const struct vslpdir *vslpd, const uint32_t key)
{
	CHECK_OBJ_NOTNULL(vslpd, VSLPDIR_MAGIC);

	const int n = vslpd->n_backend * vslpd->replicas;
	int idx = -1, high = n, low = 0, i;

        do {
            i = (high + low) >> 1;
            if (vslpd->hashcircle[i].point == key)
                idx = i;
            else if (i == n - 1)
                idx = n - 1;
            else if (vslpd->hashcircle[i].point < key && vslpd->hashcircle[i+1].point >= key)
                idx = i + 1;
            else if (vslpd->hashcircle[i].point > key)
                if (i == 0)
                    idx = 0;
                else
                    high = i;
            else
                low = i;
        } while (idx == -1);

	return idx;
}

static int //0=recoverey state, 1=healthy state
vslp_be_healthy(struct vslp_state *state, int chosen)
{
	int rval = 0;
	time_t now;

	AN(state);
	CHECK_OBJ_NOTNULL(state->vslpd, VSLPDIR_MAGIC);
	vslpdir_lock(state->vslpd);

	if(!state->vslpd->bstate[chosen].state)
	{
		now = time(NULL);
		if(state->vslpd->bstate[chosen].recover_time == 0)
			state->vslpd->bstate[chosen].recover_time = now;
		if (now >= (state->vslpd->bstate[chosen].recover_time + state->vslpd->rampup_time))
		{
			state->vslpd->bstate[chosen].state = 1;
			state->vslpd->bstate[chosen].recover_time = 0;
			rval = 1;
		}
	}
	else
		rval = 1;

	vslpdir_unlock(state->vslpd);
	return rval;
}

static void
vslp_be_unhealthy(struct vslp_state *state, int chosen)
{
	AN(state);
	CHECK_OBJ_NOTNULL(state->vslpd, VSLPDIR_MAGIC);
	vslpdir_lock(state->vslpd);

	if(state->vslpd->bstate[chosen].state)
	{
		state->vslpd->bstate[chosen].state = 0;
		state->vslpd->bstate[chosen].recover_time = 0;
	}
	vslpdir_unlock(state->vslpd);
}

static int
vslp_choose_next(struct vslp_state *state, uint32_t n_retry)
{
	int i, chosen = 0;

	AN(state);
	CHECK_OBJ_NOTNULL(state->vslpd, VSLPDIR_MAGIC);

	if(state->idx < 0 || n_retry >= state->vslpd->n_backend)
		return -1;

	do
	{
		i = state->idx + n_retry;
		if(i >= (state->vslpd->n_backend * state->vslpd->replicas))
		i = i - (state->vslpd->n_backend * state->vslpd->replicas);
	

		chosen = state->vslpd->hashcircle[i].host;
		n_retry++;
		if(n_retry > (state->vslpd->n_backend * state->vslpd->replicas) + state->vslpd->n_backend)
			return -1;
	} while (state->picklist & (1 << chosen));
	state->picklist |= 1 << chosen;

	return chosen;
}

static int
vslp_choose_next_healthy(struct vslp_state *state, uint32_t n_retry)
{
	int i, max_iter, chosen;
	VCL_BACKEND be;

	AN(state);
	CHECK_OBJ_NOTNULL(state->vslpd, VSLPDIR_MAGIC);

	max_iter = state->vslpd->n_backend - n_retry;
	for(i = 0; i < max_iter; i++, n_retry++)
	{
		chosen = vslp_choose_next(state, n_retry);
		if(chosen == -1)
			break;

		be = state->vslpd->backend[chosen];
		AN(be);

		if(be->healthy(be, NULL, NULL))
		{
			vslp_be_healthy(state, chosen);
			break;
		}
		vslp_be_unhealthy(state, chosen);
		chosen = -1;
	}

	return chosen;
}

void
vslpdir_new(struct vslpdir **vslpdp, const char *vcl_name)
{
	struct vslpdir *vslpd;

	AN(vcl_name);
	AN(vslpdp);
	AZ(*vslpdp);
	ALLOC_OBJ(vslpd, VSLPDIR_MAGIC);
	AN(vslpd);
	*vslpdp = vslpd;
	AZ(pthread_mutex_init(&vslpd->mtx, NULL));

	ALLOC_OBJ(vslpd->bstate, VSLP_BESTATE_MAGIC);
	vslpd->rampup_time = 60;
	vslpd->altsrv_p = 0.10;
	vslpd->hash_fp = vslp_hash_crc32;
}

void
vslpdir_delete(struct vslpdir **vslpdp)
{
	struct vslpdir *vslpd;

	AN(vslpdp);
	vslpd = *vslpdp;
	*vslpdp = NULL;

	CHECK_OBJ_NOTNULL(vslpd, VSLPDIR_MAGIC);

	free(vslpd->backend);
	AZ(pthread_mutex_destroy(&vslpd->mtx));
	FREE_OBJ(vslpd->bstate);
	free(vslpd->hashcircle);
	FREE_OBJ(vslpd);
}

void
vslpdir_add_backend(struct vslpdir *vslpd, VCL_BACKEND be)
{
	CHECK_OBJ_NOTNULL(vslpd, VSLPDIR_MAGIC);
	AN(be);
	vslpdir_lock(vslpd);
	if(vslpd->n_backend >= 64)
	{
		vslpdir_unlock(vslpd);
		WRONG("VSLP supports only up to 64 backends");
	}
	if(vslpd->n_backend >= vslpd->l_backend)
		vslpdir_expand(vslpd, vslpd->l_backend + 16);
	assert(vslpd->n_backend <= vslpd->l_backend);
	vslpd->backend[vslpd->n_backend] = be;
	vslpd->bstate[vslpd->n_backend].state = 1;
	vslpd->n_backend++;
	vslpdir_unlock(vslpd);
}

void
vslpdir_set_rampup_ratio(struct vslpdir *vslpd, VCL_REAL ratio)
{
	assert(ratio >= 0.0 && ratio < 1.0);
	CHECK_OBJ_NOTNULL(vslpd, VSLPDIR_MAGIC);
	vslpdir_lock(vslpd);
	vslpd->altsrv_p = ratio; 
	vslpdir_unlock(vslpd);
}

void
vslpdir_set_rampup_time(struct vslpdir *vslpd, VCL_DURATION duration)
{
	assert(duration >= 0.0);
	CHECK_OBJ_NOTNULL(vslpd, VSLPDIR_MAGIC);

	time_t t = (time_t) floor(duration);

	vslpdir_lock(vslpd);
	vslpd->rampup_time = t; 
	vslpdir_unlock(vslpd);
}

void
vslpdir_set_hash(struct vslpdir *vslpd, hash_func hash_m)
{
	CHECK_OBJ_NOTNULL(vslpd, VSLPDIR_MAGIC);
	AN(hash_m);
	vslpdir_lock(vslpd);
	vslpd->hash_fp = hash_m;
	vslpdir_unlock(vslpd);
}

void
vslpdir_init_hashcircle(struct vslpdir *vslpd, VCL_INT replicas)
{
	int i, j;

	CHECK_OBJ_NOTNULL(vslpd, VSLPDIR_MAGIC);
	AZ(vslpd->hashcircle);

	vslpdir_lock(vslpd);
	vslpd->hashcircle = calloc(vslpd->n_backend * replicas,
		sizeof(struct vslp_hostnode));
	AN(vslpd->hashcircle);
	AN(vslpd->backend);
	if(vslpd->backend[0] == NULL)
	{
		vslpdir_unlock(vslpd);
		WRONG("VSLP director doesn't have any backends");
	}
	vslpd->replicas = replicas;
		

        for (i = 0; i < vslpd->n_backend; i++) {
                for (j = 0; j < replicas; j++) {
                        int len = strlen(vslpd->backend[i]->vcl_name)
                            + (j == 0 ? 0 : log10(j)) + 2;
                        char s[len];

                        sprintf(s, "%s%d", vslpd->backend[i]->vcl_name, j);
                        vslpd->hashcircle[i * replicas + j].point =
                            vslp_hash_sha256(s);
                        vslpd->hashcircle[i * replicas + j].host = i;
                }
                vslpd->bstate[i].canon_point = vslpd->hashcircle[i * replicas].point;
        }
        qsort( (void *) vslpd->hashcircle, vslpd->n_backend * replicas,
            sizeof (struct vslp_hostnode), (compar) hostnode_compare);

        for (i = 0; i < vslpd->n_backend; i++)
	{
        	for (j = 0; j < replicas; j++)
		{
			VSL(SLT_Debug, 0, "VSLP hashcircle[%5ld] = {point = %8x, host = %2d}\n",
			    i * replicas + j,
			    vslpd->hashcircle[i * replicas + j].point,
			    vslpd->hashcircle[i * replicas + j].host);
		}
	}

	vslpdir_unlock(vslpd);
}

void
vslpdir_lock(struct vslpdir *vslpd)
{
	CHECK_OBJ_NOTNULL(vslpd, VSLPDIR_MAGIC);
	AZ(pthread_mutex_lock(&vslpd->mtx));
}

void
vslpdir_unlock(struct vslpdir *vslpd)
{
	CHECK_OBJ_NOTNULL(vslpd, VSLPDIR_MAGIC);
	AZ(pthread_mutex_unlock(&vslpd->mtx));
}

void vslpdir_expand(struct vslpdir *vslpd, unsigned n)
{
	CHECK_OBJ_NOTNULL(vslpd, VSLPDIR_MAGIC);

	vslpd->backend = realloc(vslpd->backend, n * sizeof *vslpd->backend);
	AN(vslpd->backend);
	vslpd->bstate = realloc(vslpd->bstate, n * sizeof *vslpd->bstate);
	AN(vslpd->bstate);
	vslpd->l_backend = n;
}

unsigned
vslpdir_any_healthy(struct vslpdir *vslpd)
{
	unsigned retval = 0;
        VCL_BACKEND be;
        unsigned u;

        CHECK_OBJ_NOTNULL(vslpd, VSLPDIR_MAGIC);
        vslpdir_lock(vslpd);
        for (u = 0; u < vslpd->n_backend; u++) {
		be = vslpd->backend[u];
		CHECK_OBJ_NOTNULL(be, DIRECTOR_MAGIC);
		if (be->healthy(be, NULL, NULL)) {
			retval = 1;
			break;
		}
	}
        vslpdir_unlock(vslpd);

	return (retval);
}

VCL_BACKEND vslpdir_pick_be(struct vslpdir *vslpd, const struct vrt_ctx *ctx, uint32_t hash,
		VCL_INT n_retry, VCL_BOOL altsrv_p, VCL_BOOL healthy)
{
	VCL_BACKEND be;
	int chosen, be_choice = 0, restarts_o = 0, restarts = 0;
	struct vslp_state state;

        CHECK_OBJ_NOTNULL(vslpd, VSLPDIR_MAGIC);
	CHECK_OBJ_NOTNULL(ctx, VRT_CTX_MAGIC);
	AN(ctx->vsl);

	if(altsrv_p)
		be_choice = (scalbn(random(), -31) > vslpd->altsrv_p);


	if(n_retry == 0) {
		if (ctx->bo) {
			restarts = restarts_o = ctx->bo->retries;
		} else {
			AN(ctx->req);
			restarts = restarts_o = ctx->req->restarts;
		}
	} else {
		n_retry--;
	}

	state.picklist = 0;
	state.vslpd = vslpd;
	state.ctx = ctx;

	if(vslpd->hashcircle == NULL)
		WRONG("Incomplete VSLP hashcircle");

	state.idx = vslp_get_prefered_idx(vslpd, hash);
	chosen = vslp_choose_next(&state, n_retry);

	if(chosen >= 0) {
		VSLb(ctx->vsl, SLT_Debug,
		    "VSLP picked preferred backend %2i for key %8x",
		    chosen, hash);
		be = vslpd->backend[chosen];
		AN(be);
	} else {
		be = NULL;
	}
	
	while(restarts > 0)
	{
		n_retry++;

		if(healthy)
			chosen = vslp_choose_next_healthy(&state, n_retry);
		else
			chosen = vslp_choose_next(&state, n_retry);

		if(chosen < 0) {
			VSLb(ctx->vsl, SLT_Debug,
			    "VSLP failed to find other healthy backend for key %8x in restarts: %2i/%2i - using previous",
			    hash, restarts, restarts_o);
		} else {
			VSLb(ctx->vsl, SLT_Debug,
			    "VSLP picked backend %2i for key %8x in restarts: %2i/%2i",
			    chosen, hash, restarts, restarts_o);
			be = vslpd->backend[chosen];
			AN(be);
		}

		if(--restarts <= 0)
			return be;
	}

	if (! be) {
		VSLb(ctx->vsl, SLT_Debug, "VSLP no backend found");
		return NULL;
	}

	if (! healthy) {
		AN(be);
		return (be);
	}

	if (be->healthy(be, ctx->bo, NULL))
	{
		if(!vslp_be_healthy(&state, chosen))
			 be_choice ^= be_choice;

		if(altsrv_p && !be_choice)
		{
			chosen = vslp_choose_next_healthy(&state, n_retry);
			if(chosen < 0) {
				VSLb(ctx->vsl, SLT_Debug,
				    "VSLP found no alternative backend in healthy");
			} else {
				VSLb(ctx->vsl, SLT_Debug,
				    "VSLP picked alternative backend %2i for key %8x in healthy",
				    chosen, hash);
				be = vslpd->backend[chosen];
			}
		}
	}
	else
	{
		vslp_be_unhealthy(&state, chosen);
		chosen = vslp_choose_next_healthy(&state, n_retry);
		if(chosen < 0) {
			VSLb(ctx->vsl, SLT_Debug,
			    "VSLP found no alternative backend in unhealthy");
		} else {
			VSLb(ctx->vsl, SLT_Debug,
			    "VSLP picked alternative backend %2i for key %8x in unhealthy",
			    chosen, hash);
			be = vslpd->backend[chosen];
		}
	}

	AN(be);
	return (be);
}
